File: sdks/java/connect-sdk/src/test/java/com/mobiscroll/connect/AuthTest.java
package com.mobiscroll.connect;

import static org.assertj.core.api.Assertions.assertThat;

import java.util.Arrays;

import com.mobiscroll.connect.models.AuthUrlParams;
import com.mobiscroll.connect.models.DisconnectParams;
import com.mobiscroll.connect.models.DisconnectResponse;
import com.mobiscroll.connect.models.TokenResponse;
import com.mobiscroll.connect.support.ClientFactory;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import okhttp3.mockwebserver.RecordedRequest;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class AuthTest {

    private MockWebServer server;

    @BeforeEach void setUp() throws Exception {
        server = new MockWebServer();
        server.start();
    }

    @AfterEach void tearDown() throws Exception {
        server.shutdown();
    }

    @Test void generateAuthUrlIncludesAllParams() {
        MobiscrollConnectClient client = ClientFactory.withMock(server);
        String url = client.auth().generateAuthUrl(AuthUrlParams.builder()
                .userId("user-1")
                .state("xyz")
                .scope("calendars events")
                .providers(Arrays.asList(Provider.GOOGLE, Provider.MICROSOFT))
                .build());

        assertThat(url).startsWith(server.url("/api/oauth/authorize").toString() + "?");
        assertThat(url).contains("response_type=code");
        assertThat(url).contains("client_id=test-client");
        assertThat(url).contains("redirect_uri=http");
        assertThat(url).contains("user_id=user-1");
        assertThat(url).contains("state=xyz");
        assertThat(url).contains("scope=calendars+events");
        assertThat(url).contains("providers=google");
        assertThat(url).contains("providers=microsoft");
    }

    @Test void getTokenExchangesCodeAndCachesCredentials() throws Exception {
        server.enqueue(new MockResponse().setBody(
                "{\"access_token\":\"a\",\"token_type\":\"Bearer\",\"expires_in\":3600,\"refresh_token\":\"r\"}"));
        MobiscrollConnectClient client = ClientFactory.withMock(server);

        TokenResponse t = client.auth().getToken("the-code");

        assertThat(t.getAccessToken()).isEqualTo("a");
        assertThat(client.getCredentials()).isSameAs(t);

        RecordedRequest req = server.takeRequest();
        assertThat(req.getPath()).isEqualTo("/api/oauth/token");
        assertThat(req.getMethod()).isEqualTo("POST");
        assertThat(req.getHeader("Authorization")).startsWith("Basic ");
        assertThat(req.getHeader("CLIENT_ID")).isEqualTo("test-client");
        String body = req.getBody().readUtf8();
        assertThat(body).contains("grant_type=authorization_code");
        assertThat(body).contains("code=the-code");
    }

    @Test void disconnectFallsBackToLegacyPathOn404() throws Exception {
        server.enqueue(new MockResponse().setResponseCode(404).setBody("{\"message\":\"not found\"}"));
        server.enqueue(new MockResponse().setBody("{\"success\":true}"));

        MobiscrollConnectClient client = ClientFactory.withMockAndCredentials(server);
        DisconnectResponse r = client.auth().disconnect(DisconnectParams.builder()
                .provider(Provider.GOOGLE).account("alice@example.com").build());

        assertThat(r.isSuccess()).isTrue();
        RecordedRequest first = server.takeRequest();
        assertThat(first.getMethod()).isEqualTo("POST");
        assertThat(first.getPath()).startsWith("/api/oauth/disconnect?");
        RecordedRequest second = server.takeRequest();
        assertThat(second.getMethod()).isEqualTo("POST");
        assertThat(second.getPath()).startsWith("/api/disconnect?");
        assertThat(second.getPath()).contains("provider=google");
        assertThat(second.getPath()).contains("account=alice%40example.com");
    }
}

File: sdks/go/auth_test.go
package mobiscroll_test

import (
	"context"
	"strings"
	"testing"

	mobiscroll "github.com/acidb/mobiscroll-connect-sdks/sdks/go"
	"github.com/acidb/mobiscroll-connect-sdks/sdks/go/testsupport"
)

func TestGenerateAuthURL_IncludesAllParams(t *testing.T) {
	c := mobiscroll.NewClient("the-client", "secret", "https://app/cb",
		mobiscroll.WithBaseURL("https://example.test/api"),
	)
	url := c.Auth().GenerateAuthURL(&mobiscroll.AuthURLParams{
		UserID:    "user-1",
		State:     "xyz",
		Scope:     "calendars events",
		Providers: []mobiscroll.Provider{mobiscroll.ProviderGoogle, mobiscroll.ProviderMicrosoft},
	})
	if !strings.HasPrefix(url, "https://example.test/api/oauth/authorize?") {
		t.Fatalf("unexpected URL prefix: %s", url)
	}
	for _, want := range []string{
		"response_type=code",
		"client_id=the-client",
		"redirect_uri=https%3A%2F%2Fapp%2Fcb",
		"user_id=user-1",
		"state=xyz",
		"scope=calendars+events",
		"providers=google",
		"providers=microsoft",
	} {
		if !strings.Contains(url, want) {
			t.Errorf("URL missing %q: %s", want, url)
		}
	}
}

func TestGetToken_ExchangesCodeAndSetsCredentials(t *testing.T) {
	srv := testsupport.NewMockServer(t)
	srv.EnqueueJSON(`{"access_token":"a","token_type":"Bearer","expires_in":3600,"refresh_token":"r"}`)

	c := mobiscroll.NewClient("the-client", "secret", "https://app/cb",
		mobiscroll.WithBaseURL(srv.URL+"/api"),
	)
	tok, err := c.Auth().GetToken(context.Background(), "the-code")
	if err != nil {
		t.Fatalf("GetToken: %v", err)
	}
	if tok.AccessToken != "a" || tok.RefreshToken != "r" {
		t.Fatalf("unexpected tokens: %+v", tok)
	}
	if got := c.Credentials(); got == nil || got.AccessToken != "a" {
		t.Fatalf("credentials not stored: %+v", got)
	}

	reqs := srv.Requests()
	if len(reqs) != 1 {
		t.Fatalf("expected 1 request, got %d", len(reqs))
	}
	r := reqs[0]
	if r.Method != "POST" || r.Path != "/api/oauth/token" {
		t.Fatalf("unexpected request: %s %s", r.Method, r.Path)
	}
	if !strings.HasPrefix(r.Header.Get("Authorization"), "Basic ") {
		t.Errorf("expected Basic auth header, got %q", r.Header.Get("Authorization"))
	}
	if r.Header.Get("CLIENT_ID") != "the-client" {
		t.Errorf("expected CLIENT_ID header, got %q", r.Header.Get("CLIENT_ID"))
	}
	body := string(r.Body)
	for _, want := range []string{"grant_type=authorization_code", "code=the-code"} {
		if !strings.Contains(body, want) {
			t.Errorf("body missing %q: %s", want, body)
		}
	}
}

func TestDisconnect_FallsBackToLegacyPathOn404(t *testing.T) {
	srv := testsupport.NewMockServer(t)
	srv.Enqueue(testsupport.MockResponse{Status: 404, Body: `{"message":"not found"}`,
		Headers: map[string]string{"Content-Type": "application/json"}})
	srv.EnqueueJSON(`{"success":true}`)

	c := mobiscroll.NewClient("id", "secret", "https://app/cb",
		mobiscroll.WithBaseURL(srv.URL+"/api"),
	)
	c.SetCredentials(&mobiscroll.TokenResponse{AccessToken: "at", RefreshToken: "rt"})

	resp, err := c.Auth().Disconnect(context.Background(), &mobiscroll.DisconnectParams{
		Provider: mobiscroll.ProviderGoogle,
		Account:  "alice@example.com",
	})
	if err != nil {
		t.Fatalf("Disconnect: %v", err)
	}
	if !resp.Success {
		t.Fatalf("expected success=true")
	}

	reqs := srv.Requests()
	if len(reqs) != 2 {
		t.Fatalf("expected 2 requests, got %d", len(reqs))
	}
	if reqs[0].Method != "POST" || !strings.HasPrefix(reqs[0].Path, "/api/oauth/disconnect?") {
		t.Errorf("first request: %s %s", reqs[0].Method, reqs[0].Path)
	}
	if reqs[1].Method != "POST" || !strings.HasPrefix(reqs[1].Path, "/api/disconnect?") {
		t.Errorf("second request: %s %s", reqs[1].Method, reqs[1].Path)
	}
	for _, want := range []string{"provider=google", "account=alice%40example.com"} {
		if !strings.Contains(reqs[1].Path, want) {
			t.Errorf("legacy path missing %q: %s", want, reqs[1].Path)
		}
	}
}

func TestGetConnectionStatus_FallsBackToLegacyPathOn404(t *testing.T) {
	srv := testsupport.NewMockServer(t)
	srv.Enqueue(testsupport.MockResponse{Status: 404, Body: `{"message":"not found"}`,
		Headers: map[string]string{"Content-Type": "application/json"}})
	srv.EnqueueJSON(`{"connections":{"google":[{"id":"u@g.com"}]},"limitReached":false}`)

	c := mobiscroll.NewClient("id", "secret", "https://app/cb",
		mobiscroll.WithBaseURL(srv.URL+"/api"),
	)
	c.SetCredentials(&mobiscroll.TokenResponse{AccessToken: "at", RefreshToken: "rt"})

	status, err := c.Auth().GetConnectionStatus(context.Background())
	if err != nil {
		t.Fatalf("GetConnectionStatus: %v", err)
	}
	if status.LimitReached {
		t.Fatalf("expected limitReached=false")
	}
	if accs, ok := status.Connections[mobiscroll.ProviderGoogle]; !ok || len(accs) != 1 || accs[0].ID != "u@g.com" {
		t.Fatalf("unexpected connections: %+v", status.Connections)
	}
}
